### PR TITLE
Marby local

### PR DESCRIPTION
**Pull request by marby3**
I'll add the keyboard I made.
I made a keyboard that uses Joystick and rotary encoder.

## Description
formy46

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/formy46/formy46.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
+        4, 4, 4, 4, 4, 4
+} };
+#endif
+
+void matrix_init_kb(void) {
+	matrix_init_user();
+};
```
Comment (drashna) [inline keyboards/formy46/formy46.c:55]:
```suggestion
```

File: keyboards/formy46/info.json [hunk 2]
```diff
@@ -0,0 +1,59 @@
+{
+    "keyboard_name": "formy46",
+    "url": "",
+    "maintainer": "marby",
+    "width": 11.5,
+    "height": 5.75,
```
Comment (drashna) [inline keyboards/formy46/info.json:6]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/formy46/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+                    layer_move(_GAME_MODE);
+                }
+            } else {
+                if (clockwise) {
+                    //Redo
+                    register_code(KC_LCTL);
+                    register_code(KC_LSFT);
+                    tap_code(KC_Z);
+                    unregister_code(KC_LCTL);
+                    unregister_code(KC_LSFT);
```
Comment (drashna) [inline keyboards/formy46/keymaps/default/keymap.c:207]:
```suggestion
                    tap_code16(C(S(KC_Z)));
```

File: keyboards/formy46/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+                } else {
+                    //Undo
+                    register_code(KC_LCTL);
+                    tap_code(KC_Z);
+                    unregister_code(KC_LCTRL);
```
Comment (drashna) [inline keyboards/formy46/keymaps/default/keymap.c:212]:
```suggestion
                    tap_code(C(KC_Z));
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/formy46/keymaps/default/rules.mk [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+SPLIT_KEYBOARD          =  yes       # Enable Split Keyboard
+RGB_MATRIX_ENABLE       =  yes       # Enable keyboard RGB Matrix
+RGB_MATRIX_DRIVER       =  WS2812    # RGB Matrix driver
+SRC                    +=  analog.c  # adc driver
+JOYSTICK_ENABLE         =  yes       # Enable Joystick
+POINTING_DEVICE_ENABLE  =  yes       # Enable Pointing Device
+ENCODER_ENABLE          =  yes       # Enable Rotary Encoder
+OLED_ENABLE             =  yes       # Enable OLED
+OLED_DRIVER             =  SSD1306   # OLED Type
+MOUSEKEY_ENABLE  		=  yes       # Enable Mouse keycord
```
Comment (drashna) [inline keyboards/formy46/keymaps/default/rules.mk:11]: These should be defined at the keyboard level, and not the keymap level.

File: keyboards/formy46/rules.mk [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+MCU  =  atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (drashna) [inline keyboards/formy46/rules.mk:11]:
```suggestion
```

File: keyboards/tyxos63/config.h [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+/*== or choose animations ==*/
+//#    define RGBLIGHT_EFFECT_BREATHING
+//#    define RGBLIGHT_EFFECT_RAINBOW_MOOD
+//#    define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+//#    define RGBLIGHT_EFFECT_SNAKE
+//#    define RGBLIGHT_EFFECT_KNIGHT
+//#    define RGBLIGHT_EFFECT_CHRISTMAS
+//#    define RGBLIGHT_EFFECT_STATIC_GRADIENT
+//#    define RGBLIGHT_EFFECT_RGB_TEST
+//#    define RGBLIGHT_EFFECT_ALTERNATING
```
Comment (drashna) [inline keyboards/tyxos63/config.h:59]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/tyxos63/info.json [hunk 8]
```diff
@@ -0,0 +1,75 @@
+{
+    "keyboard_name": "tyxos63",
+    "url": "",
+    "maintainer": "marby3",
+    "width": 16,
+    "height": 5.25,
```
Comment (drashna) [inline keyboards/tyxos63/info.json:6]:
```suggestion
```

File: keyboards/tyxos63/matrix.c [hunk 9]
```diff
@@ -0,0 +1,284 @@
+/*
```
Comment (drashna) [inline keyboards/tyxos63/matrix.c:1]: Please convert this to matrix lite. 

https://docs.qmk.fm/#/custom_matrix?id=lite

File: keyboards/tyxos63/rules.mk [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (drashna) [inline keyboards/tyxos63/rules.mk:11]:
```suggestion
```

File: keyboards/tyxos63/rules.mk [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+BOOTLOADER = caterina
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
```
Comment (drashna) [inline keyboards/tyxos63/rules.mk:17]: There is no "full" bootmagic anymore, only the "lite" option, which can be turned on with just "yes".
```suggestion
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
```

File: keyboards/tyxos63/rules.mk [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
```
Comment (drashna) [inline keyboards/tyxos63/rules.mk:28]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by marby3** [comment]
Sorry, the data from my other keyboard was mixed in.
I have deleted it.

File: keyboards/formy46/keymaps/default/rules.mk (hunk 5 above)
Comment (marby3) [reply keyboards/formy46/keymaps/default/rules.mk:11]: I moved as much as I could to the keyboard.

File: keyboards/formy46/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+            break;
+        case _FN:
+            oled_write_P(PSTR("Main\n"), false);
+            oled_write_P(PSTR("Game\n"), false);
+            oled_write_P(PSTR("Crea\n"), false);
+            oled_write_P(PSTR("Func\n"), true);
+            break;
+        default:
+            oled_write_ln_P(PSTR("Undefined\n"), false);
+    }
```
[the remark this answers is not in the document]
Comment (marby3) [reply keyboards/formy46/keymaps/default/keymap.c]: I'll use it as is!

File: keyboards/formy46/rules.mk [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+RGBLIGHT_ENABLE         =  no
+RGB_MATRIX_ENABLE       =  no       # Enable keyboard RGB Matrix
+RGB_MATRIX_DRIVER       =  WS2812    # RGB Matrix driver
+SRC                    +=  analog.c  # adc driver
+JOYSTICK_ENABLE         =  yes       # Enable Joystick
+POINTING_DEVICE_ENABLE  =  yes       # Enable Pointing Device
+ENCODER_ENABLE          =  yes       # Enable Rotary Encoder
+OLED_ENABLE             =  yes       # Enable OLED
+OLED_DRIVER             =  SSD1306   # OLED Type
+MOUSEKEY_ENABLE  		=  yes       # Enable Mouse keycord
```
Comment (drashna) [inline keyboards/formy46/rules.mk:34]:
```suggestion
BOOTMAGIC_ENABLE        =  no        # Enable Bootmagic Lite
MOUSEKEY_ENABLE         =  yes       # Mouse keys
EXTRAKEY_ENABLE         =  no        # Audio control and System control
CONSOLE_ENABLE          =  no        # Console for debug
COMMAND_ENABLE          =  no        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE        =  no        # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE             =  no        # USB Nkey Rollover
BACKLIGHT_ENABLE        =  no        # Enable keyboard backlight functionality
RGBLIGHT_ENABLE         =  no        # Enable keyboard RGB underglow
AUDIO_ENABLE            =  no        # Audio output
LTO_ENABLE              =  yes       # Reduced firmware size

# Formy46 Features Options
SPLIT_KEYBOARD          =  yes       # Enable Split Keyboard
RGB_MATRIX_DRIVER       =  WS2812    # RGB Matrix driver
SRC                    +=  analog.c  # adc driver
JOYSTICK_ENABLE         =  yes       # Enable Joystick
POINTING_DEVICE_ENABLE  =  yes       # Enable Pointing Device
ENCODER_ENABLE          =  yes       # Enable Rotary Encoder
OLED_ENABLE             =  yes       # Enable OLED
OLED_DRIVER             =  SSD1306   # OLED Type
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/formy46/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+    #ifdef RGB_MATRIX_ENABLE
+        oled_write_ln_P(PSTR(""), false);
+        oled_write_ln_P(PSTR("LED:"), false);
+        oled_write_ln_P(PSTR(""), false);
+        static char rgbMode[6] = {0};
+        snprintf(rgbMode, sizeof(rgbMode), "M:%-3d", rgblight_get_mode());
```
[the remark this answers is not in the document]
Comment (marby3) [reply keyboards/formy46/keymaps/default/keymap.c:286]: I used the code above as a reference, but it did not reduce the firmware size!
Maybe I am doing something wrong....

`        char     rgblight_mode[1];
        uint16_t rgb_mode     = rgblight_get_mode();
        rgblight_mode[0] = rgb_mode;
        oled_write_P(PSTR("M:"), false);
        oled_write(rgblight_mode, false);
        char     rgblight_hue[1];
        uint16_t rgb_hue     = rgblight_get_hue();
        rgblight_hue[0] = rgb_hue;
        oled_write_P(PSTR("H:"), false);
        oled_write(rgblight_hue, false);
        char     rgblight_sat[1];
        uint16_t rgb_sat     = rgblight_get_sat();
        rgblight_sat[0] = rgb_sat;
        oled_write_P(PSTR("S:"), false);
        oled_write(rgblight_sat, false);
        char     rgblight_val[1];
        uint16_t rgb_val     = rgblight_get_val();
        rgblight_val[0] = rgb_val;
        oled_write_P(PSTR("V:"), false);
        oled_write(rgblight_val, false);`

File: keyboards/formy46/info.json [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+                {"x":0, "y":3.25},
+                {"x":1, "y":3.25},
+                {"x":12.5, "y":3.25},
+                {"x":13.5, "y":3.25},
+                {"x":4, "y":4},
+                {"x":5, "y":4},
+                {"x":6, "y":4},
+                {"x":7.5, "y":4},
+                {"x":8.5, "y":4},
+                {"x":9.5, "y":4}
```
Comment (noroadsleft) [inline keyboards/formy46/info.json:53]:
```suggestion
                {"x":0, "y":0.25},
                {"x":1, "y":0.25},
                {"x":2, "y":0},
                {"x":3, "y":0},
                {"x":4, "y":0.25},
                {"x":9.5, "y":0.25},
                {"x":10.5, "y":0},
                {"x":11.5, "y":0},
                {"x":12.5, "y":0.25},
                {"x":13.5, "y":0.25},

                {"x":0, "y":1.25},
                {"x":1, "y":1.25},
                {"x":2, "y":1},
                {"x":3, "y":1},
                {"x":4, "y":1.25},
                {"x":9.5, "y":1.25},
                {"x":10.5, "y":1},
                {"x":11.5, "y":1},
                {"x":12.5, "y":1.25},
                {"x":13.5, "y":1.25},

                {"x":0, "y":2.25},
                {"x":1, "y":2.25},
                {"x":2, "y":2},
                {"x":3, "y":2},
                {"x":4, "y":2.25},
                {"x":5, "y":3},
                {"x":8.5, "y":3},
                {"x":9.5, "y":2.25},
                {"x":10.5, "y":2},
                {"x":11.5, "y":2},
                {"x":12.5, "y":2.25},
                {"x":13.5, "y":2.25},

                {"x":0, "y":3.25},
                {"x":1, "y":3.25},
                {"x":2, "y":3},
                {"x":3, "y":3},
                {"x":4, "y":4},
                {"x":5, "y":4},
                {"x":6, "y":4},
                {"x":7.5, "y":4},
                {"x":8.5, "y":4},
                {"x":9.5, "y":4},
                {"x":10.5, "y":3},
                {"x":11.5, "y":3},
                {"x":12.5, "y":3.25},
                {"x":13.5, "y":3.25}
```

File: keyboards/formy46/readme.md [hunk 17]
```diff
@@ -0,0 +1,35 @@
+# formy46
+===
+[formy42](https://imgur.com/a/a3PWhZx)
```
Comment (noroadsleft) [inline keyboards/formy46/readme.md:3]:
```suggestion
# formy46

[formy42](https://imgur.com/a/a3PWhZx)
```

File: keyboards/formy46/readme.md [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+## Description
+A compact and multifunctional split keyboard.
+
+* Keyboard Maintainer: [marby3](https://github.com/yourusername)
+* Hardware Supported: *Formy46PCB, ProMicro, MX Swhitch, Choc Switch*
```
Comment (noroadsleft) [inline keyboards/formy46/readme.md:9]:
```suggestion
* Hardware Supported: *Formy46PCB, ProMicro, MX Switch, Choc Switch*
```

File: keyboards/formy46/rules.mk [hunk 19]
```diff
@@ -0,0 +1,34 @@
+# MCU name
+MCU  =  atmega32u4
+
+# Bootloader selection
+BOOTLOADER  =  caterina
+
+# Build Options
+BOOTMAGIC_ENABLE        =  no         # Virtual DIP switch configuration
+MOUSEKEY_ENABLE         =  no        # Mouse keys
```
Comment (noroadsleft) [inline keyboards/formy46/rules.mk:9]: You have this defined twice.

```suggestion
MOUSEKEY_ENABLE         =  yes       # Mouse keys
```

File: keyboards/formy46/rules.mk (hunk 14 above)
Comment (noroadsleft) [inline keyboards/formy46/rules.mk:34]:
```suggestion
```

File: keyboards/formy46/rules.mk [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+EXTRAKEY_ENABLE         =  no        # Audio control and System control
+CONSOLE_ENABLE          =  no        # Console for debug
+COMMAND_ENABLE          =  no        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE        =  no        # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE             =  no        # USB Nkey Rollover
+BACKLIGHT_ENABLE        =  no        # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE         =  no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE        =  no        # Enable Bluetooth
```
Comment (noroadsleft) [inline keyboards/formy46/rules.mk:19]:
```suggestion
```

File: keyboards/formy46/rules.mk [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+AUDIO_ENABLE            =  no        # Audio output
+LTO_ENABLE              =  yes       # Reduced firmware size
+
+# Formy46 Features Options
+SPLIT_KEYBOARD          =  yes       # Enable Split Keyboard
+RGBLIGHT_ENABLE         =  no
```
Comment (noroadsleft) [inline keyboards/formy46/rules.mk:25]: Also defined twice.

```suggestion
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/formy46/keymaps/default/keymap.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (drashna) [reply keyboards/formy46/keymaps/default/keymap.c:287]: if you are using snprintf elsewhere, then it wont.

**Comment by marby3** [comment]
I'm sorry, but the layout has changed, so I'm going to close this once.

**Pull request closed without merging**